File: ios/SafeBox/Disguise/DisguiseRegistry.swift
import Foundation

/// The compiled-in, ordered, **append-only** face list (decisions §1.6). A
/// shipped face is never removed: an install enrolled on it would otherwise be
/// unable to render its own lock screen.
@MainActor
struct DisguiseRegistry {
    /// Fail-closed target for every unresolvable id (§3). Deliberately
    /// `nonisolated` so the envelope decoder — which runs off the main actor —
    /// can name it.
    nonisolated static let defaultId = "calculator"

    let all: [any DisguiseProviding]

    init(all: [any DisguiseProviding] = [CalculatorDisguise(), NumpadDisguise(), PatternDisguise()]) {
        self.all = all
    }

    /// The default face. The registry is never empty (the initializer's default
    /// argument is the shipped list), but a caller-supplied empty list still
    /// resolves rather than trapping.
    var defaultDisguise: any DisguiseProviding {
        all.first { $0.id == Self.defaultId } ?? all[0]
    }

    /// Fail closed: missing, unknown, or undecodable ids all render the
    /// calculator. Never an error surface, never a non-disguise surface.
    func resolve(id: String?) -> any DisguiseProviding {
        guard let id, let match = all.first(where: { $0.id == id }) else {
            return defaultDisguise
        }
        return match
    }
}

File: android/app/src/main/java/com/calcplus/calculator/core/disguise/DisguiseRegistry.kt
package com.calcplus.calculator.core.disguise

/**
 * The compiled-in list of lock faces, in registry order (§1.6).
 *
 * Append-only: a shipped face is never removed, because an install enrolled on
 * it would otherwise become unopenable (skeleton §10 risk 4). The first entry
 * is the default and the fail-closed target.
 */
class DisguiseRegistry(val faces: List<DisguiseProvider>) {
    init {
        require(faces.isNotEmpty()) { "the registry needs at least the default face" }
        require(faces.map { it.id }.toSet().size == faces.size) { "disguise ids must be unique" }
    }

    /** Calculator — what an unresolvable id renders as, in `disguise` mode. */
    val default: DisguiseProvider = faces.first()

    /**
     * Fail closed: a missing, unknown or undecodable id resolves to [default].
     * Never an error surface, never a non-disguise surface (§3).
     */
    fun resolve(id: String?): DisguiseProvider =
        faces.firstOrNull { it.id == id } ?: default
}
